packages/api-key/src/services/api-key-module-service.ts

```diff
@@ -79,13 +86,42 @@ export default class ApiKeyModuleService<TEntity extends ApiKey = ApiKey>
     data: ApiKeyTypes.CreateApiKeyDTO | ApiKeyTypes.CreateApiKeyDTO[],
     @MedusaContext() sharedContext: Context = {}
   ): Promise<TEntity | TEntity[]> {
-    const data_ = Array.isArray(data) ? data : [data]
+    const asArray = Array.isArray(data) ? data : [data]
+    await this.validateCreateApiKeys(asArray, sharedContext)
+
+    const normalizedInput: CreateApiKeyDTO[] = []
+    const generatedTokens: TokenDTO[] = []
+    for (const key of asArray) {
+      let tokenData: TokenDTO
+      if (key.type === "publishable") {
+        tokenData = ApiKeyModuleService.generatePublishableKey()
+      } else {
+        tokenData = await ApiKeyModuleService.generateSecretKey()
+      }
+
+      generatedTokens.push(tokenData)
+      normalizedInput.push({
+        ...key,
+        token: tokenData.hashedToken,
+        salt: tokenData.salt,
+        redacted: tokenData.redacted,
+      })
+    }
 
     const createdApiKeys = await this.apiKeyService_.create(
-      data_,
+      normalizedInput,
       sharedContext
     )
 
+    // TODO: This seems to affect what is ultimately written in the DB, fix.
```

Comment:
@adrien2p I wanted to discuss this with you. Unfortunately, microORM depends on some non-enumerable fields from the object it returns for serialization, so we have to mutate here.

However, due to the unit of work approach it follows, mutating the data here means that what actually gets written in the DB is the raw token, and not the hashed token as expected.

What do you think the best approach would be here (and similarly, below in retrievals)? I thought of doing is:
1. Do the token reassigning after the serialization happens
2. Something else you are aware of? :D

Reply:
based on our discussion, the best approach here would be for the protected create method to return a map (or similar) with the api keys and the raw token and perform the assign after serialization